player/weapons/revolver/revolver.gd

```diff
@@ -0,0 +1,75 @@
+class_name Revolver
+extends Weapon
+
+const Scene = preload("res://player/weapons/revolver/revolver.tscn")
+
+var damage: float = 1.0
+var base_fire_rate: float = 0.75
+var reload_time: float = 0.35
+
+var _magazine: PlayerMagazine
+var _is_reloading: bool = false
+var _is_shooting: bool = false
+
+@onready var shoot_audio_stream: AudioStreamPlayer2D = $ShootAudioStream
+@onready var no_ammo_audio_stream: AudioStreamPlayer2D = $NoAmmoAudioStream
+@onready var reload_audio_stream: AudioStreamPlayer2D = $ReloadAudioStream
+@onready var shoot_area: Area2D = $ShootArea
+
+
+static func create() -> Revolver:
+	var instance = Scene.instantiate()
+	return instance
+
+
+func _ready() -> void:
+	_magazine = PlayerMagazine.new(7)
+
+
+func start_shooting(point: Vector2) -> void:
+	if not _magazine.has_ammo():
+		no_ammo_audio_stream.play()
+		return
+
+	if _is_shooting:
+		return
+
+	_is_reloading = false
+	_is_shooting = true
+
+	shoot_audio_stream.play()
+	_magazine.unload(1)
+
+	shoot_area.global_position = point
+
+	var enemies = shoot_area.get_overlapping_bodies()
+	enemies.sort_custom(y_sort)
+	var enemy = enemies.front()
+
+	if enemy:
+		enemy.hit.emit(damage)
+	else:
+		SignalBus.shot_hit_ground.emit(point)
+
+	await get_tree().create_timer(base_fire_rate).timeout
```

Comment:
i'm still getting used to this approach - it's pretty straight forward but idk how easy it is to add more logic to. once you're here `await`ing, nothing else can happen

having things tracked with ticks and/or utilizing delta times would probably be around the same amount of code but then you can just put things into the `_process`. or for doing things like `get_overlapping_bodies` that would be done in the `_physics_process`.

a simple state machine would work here as well

Reply:
> I'm not sure I follow how doing these things can improve if we move them into process, in my eyes it's mixing two different things

yeah i mean, when you await something in this function, there's nothing you can do to stop it or change behavior vs having per-frame logic to make decisions on

unless you then check for other conditions after the await, you're pretty much committed to
doing whatever comes after it no matter what, which seems harder to expand upon. in this case, after the await it's just setting _is_shooting to false but if other things are going on around this method call, they could contradict themselves.